### PR TITLE
feat(custom-date-picker): updated custom date, time, date range pickers w flatpickr

### PR DESCRIPTION
**Pull request by wbarbee**
## Summary

Refactors **datepicker**, **timepicker**, **date range picker** components, utilizing [`flatpickr`](https://flatpickr.js.org/) third-party library. 

## ADO Story or GitHub Issue Link

[UI: Datepicker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910721)
[UI: Date/time range picker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910724)
[UI: Timepicker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910722)
[UI: Flatpickr theme User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1937059)

## Figma Link
[https://www.figma.com/design/dhPuQQrqxHqMvtmnMMuTry/Santorini---Global-Filters?node-id=518-47488&node-type=frame&m=dev](https://www.figma.com/design/dhPuQQrqxHqMvtmnMMuTry/Santorini---Global-Filters?node-id=518-47488&node-type=frame&m=dev)

## Notes

**primary considerations**:
- validation -- flatpickr out of the box/plugins vs. dropping all validations except for `invalidText`
- custom shidoka theming

## To Do

- potential 2.0 breaking changes **chore** -- restructure Storybook side nav items for these three components, possibly nesting them under a common folder (e.g., `dateTimePicker`, `Date Time Picker`)

## Checklist

- [x] Used Conventional Commit messages as outlined in the contributing guide.
  - [x] Noted breaking changes (if any).
- [x] Documented/updated all props, events, slots, parts, etc with JSDoc.
  - [x] Ran the `analyze` command to update Storybook docs.
- [x] Added/updated Stories with controls to cover all variants.
- [x] Ran `test` locally to address any failures.
- [x] Added/updated the Figma link for the Story's Design tab.
- [x] Added any new component exports to the src/index.ts file

## Screenshots

Screenshot 2024-10-16 at 9 45 04 AM

Screenshot 2024-10-16 at 4 51 30 PM

Screenshot 2024-10-16 at 9 45 51 AM

Screenshot 2024-10-16 at 9 46 54 AM

### CODE REVIEW

File: .gitignore
Comment (wbarbee) [inline .gitignore:12]: this is a macOS thing. i still don't fully understand what triggers it being included in a commit, but adding to the .gitignore as it does not include any project-relevant or project-critical info

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:107]: Should this be a calendar icon, why overflow?

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:148]: A button anchor concept is interesting, but I don't understand how it can function correctly with a form if there is no `<input>`, and how the user is supposed to know the selected date without opening the picker.

File: src/components/reusable/daterangepicker/daterangepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/daterangepicker/daterangepicker.stories.js:246]: I think we just need to go with one variant, probably single input.

**Review by brian-patrick-3 — COMMENTED** [review]
There's a lot going on in here, so I just left a few high level comments on things that jumped out at me for now.

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:107]: for this comment and your 3rd comment about the overflow icon, i was working off of the implementation that they have in the figma file where it would be part of a global filter or something.

i suppose we could just have two options -- input or button, but i originally had it slotted instead of internal for flexibility of the anchor. let me know what you think re: having 2-3 hardcoded internal options, just input, etc.

Comment (brian-patrick-3) [reply src/components/reusable/datePicker/datepicker.stories.js:107]: I see that in the old design now. I think what the design is getting at there, is they just want some sort of popup menu to house the datepicker in that scenario. I think that design is not well thought out. The calendar selection should probably still be tied to an actual input, and that's how all of the flatpickr examples work. Even the [inline calendar example](https://flatpickr.js.org/examples/#inline-calendar) still shows an input, though maybe there could be a way to hide it.

I think the most important thing is that everything still works with forms. Slotted elements are also exposed to external styling, since they live in the light dom, something to consider.

Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:107]: understood. yeah it's been a little difficult to get the design nailed down on this one. still working to get better about communicating the need to have things more clearly defined. 

in the meantime, i've tried to pack in features that i can rip out when the time comes, which is overkill. but easy to scale it back to more internal, simpler variations.

i'll take another stab with all of this in mind and see if we can get it closer

Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:107]: oh, yes this should be a calendar icon

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js]: this has been fixed by moving majority of slots internal, minus the label unnamed slot

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js]: fixed this in migrating inputs from slot to internal

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js]: button options removed in datepicker, timepicker, daterangepicker, starting with an internal, single input element that fills 100% of parent/containing element

i can implement sizing (sm, md, lg) if we want to do that in this case.

File: src/components/reusable/daterangepicker/daterangepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/daterangepicker/daterangepicker.stories.js]: ✅

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js]: ok this should be a bit more on track. i've set `required` to true in storybook for each component by default, as that's what initializes the validation and it will be easier to test. 

two behaviors that i've noticed with flatpickr:
1) for timepicker, if you click to open the timepicker and do not 'touch it' or modify the time at all, then close it by clicking anywhere outside of the overlay, there will be a moment where it does my standard validation and recognizes that the field is required, that no time has been chosen. however, i think flatpicker then defaults to 12:00

2) on the daterangepicker, when you select only one date, my validations are working as expected. however, when you open the datepicker overlay again, it has reset to no dates selected. this is more of a standard date picker behavior, but just wanted to note that as well.

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:69]: Part of the 2.0 changes has some [chores for the form components](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1931303/), can you please include those changes to these components as well? I removed them from the tasks for that story so you don't get conflicts.

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:57]: I'm not seeing the custom error message appear under the component when I update this control.

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:191]: When I click on the caption or the error message, it also opens the picker.

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:192]: Caption styles don't seem to match the other form components, it should be common through the form-input scss

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:56]: When I add warning text is appears twice.

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.stories.js:54]: If you update the `patterns/forms` page with these components set as required, it will be easier to test the validation with form submit.

File: src/common/scss/form-input.scss
Comment (brian-patrick-3) [inline src/common/scss/form-input.scss:84]: Shouldn't need to specify this font size, it's the default body font size.

File: src/components/reusable/datePicker/datepicker.scss
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.scss:8]: We don't set a default width on any of the other form inputs, usually just let the application control it.

File: src/components/reusable/timepicker/timepicker.stories.js
Comment (brian-patrick-3) [inline src/components/reusable/timepicker/timepicker.stories.js:41]: When I set 12:00 AM on the timepicker, it appears as `00:00`

Comment (wbarbee) [reply src/components/reusable/timepicker/timepicker.stories.js:41]: this is a flatpickr thing, i think. you can test here:

https://flatpickr.js.org/examples/#time-picker

Comment (brian-patrick-3) [reply src/components/reusable/timepicker/timepicker.stories.js:41]: I see, interesting default. I wonder if we can [change the default format](https://flatpickr.js.org/formatting/#time-formatting-tokens) so it shows as "12:00 AM"? Actually, I don't see a prop on the timepicker component to modify the format like datepicker has.

Comment (wbarbee) [reply src/components/reusable/timepicker/timepicker.stories.js:41]: from what i could tell from the docs, the dateFormat doesn't behave quite the same for Timepicker as it does for Datepicker. i tried another time format like `H:i:s` and it didn't work.

that's why i'm hardcoding that value, but i can try again to make it work and get the 12H format to show in the input

File: src/components/reusable/datePicker/datepicker.scss
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.scss]: ✅

File: src/common/scss/form-input.scss
Comment (wbarbee) [reply src/common/scss/form-input.scss]: ✅

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:59]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts]: ✅

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js]: ✅

File: src/components/reusable/timepicker/timepicker.stories.js
Comment (brian-patrick-3) [reply src/components/reusable/timepicker/timepicker.stories.js:44]: I wonder if something like `h:i K` is what we want there for the 12 hour format default.

Comment (wbarbee) [reply src/components/reusable/timepicker/timepicker.stories.js:44]: that works, nice 👍

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:60]: ✅

File: src/components/reusable/datePicker/datepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.stories.js:57]: ✅

File: src/components/reusable/timepicker/timepicker.stories.js
Comment (wbarbee) [reply src/components/reusable/timepicker/timepicker.stories.js:44]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:189]: this one is a bit elusive. i don't know if this is a weird side-effect of flatpickr as well. i've tried adjusting the html and pointer-events but not currently working as expected. still working on it

Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:207]: ✅

**Review by brian-patrick-3 — COMMENTED** [review]
A couple more things:

1. The validations don't appear to be working with the form, it's treating the form as valid even without entering values for any of the date/time inputs. Take a look at the console on the forms page when clicking submit.
   1. I think you are just missing some of the pieces in the `_validate` function that the other form components have, ex setting validity on the `_internals`.
2. I don't see any migration notes yet. I'm curious what exactly the breaking changes are. Ex: is there a new format for `value`? is locale is no longer automatic? etc

**Comment by wbarbee** [comment]
> A couple more things:
> 
> 1. The validations don't appear to be working with the form, it's treating the form as valid even without entering values for any of the date/time inputs. Take a look at the console on the forms page when clicking submit.
>    
>    1. I think you are just missing some of the pieces in the `_validate` function that the other form components have, ex setting validity on the `_internals`.
> 2. I don't see any migration notes yet. I'm curious what exactly the breaking changes are. Ex: is there a new format for `value`? is locale is no longer automatic? removed/changed props? etc

i amended a comment above. still having some issues with the validations, but it's nearly there. aiming to get it sorted tomorrow (mon 21st)

File: src/stories/migration.mdx
Comment (brian-patrick-3) [inline src/stories/migration.mdx:31]: Should this be "can" or "must"?

File: src/stories/migration.mdx
Comment (brian-patrick-3) [inline src/stories/migration.mdx:32]: This is visual output right? Should we mention anything about the code input/output formats?

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:28]: What happened to `size`?

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:48]: Should this be `name` to remain consistent with other form components?

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:269]: We shouldn't need an aria-label if there is visible text. aria-label is really just for interactive elements with no visible text.

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:252]: This aria-label is probably only needed on the icon, since the div also has visible text.

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:396]: We should mention the event output structure has changed.

File: src/components/reusable/datePicker/datepicker.ts
Comment (brian-patrick-3) [inline src/components/reusable/datePicker/datepicker.ts:395]: I notice this is a string for a single date, but an array of Date objects for a multi-select. Should the type be consistent (string vs date obj)?

**Review by brian-patrick-3 — COMMENTED** [review]
I notice some console errors on load from the timepicker and daterangepicker

File: src/stories/migration.mdx
Comment (wbarbee) [reply src/stories/migration.mdx]: ✅ -- totally reorganized this migration guide to be more of a checklist for devs

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:28]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:274]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts]: ✅

File: src/stories/migration.mdx
Comment (wbarbee) [reply src/stories/migration.mdx]: ✅

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:400]: corrected -- handling this as a Date object or an array of Date objects, depending on single vs. multiple `mode`.

File: src/components/reusable/datePicker/datepicker.ts
Comment (wbarbee) [reply src/components/reusable/datePicker/datepicker.ts:401]: ✅

**Review by brian-patrick-3 — APPROVED** [review]

**Comment by brian-patrick-3** [comment]
LGTM! Once QA approves, you can squash merge and include the `BREAKING CHANGE:` footer notes in the extended description textarea.

**Comment by wbarbee** [comment]
fixed the following:

1) minDate / maxDate: when these values are set to the same year (ex: 2024), Flatpickr hides the year from the date header in each calendar
**_FIX_:**  we are now detecting this and removing year element entirely when minDate and maxDate are explicitly set and are of the same year -- otherwise, the year will show as normal
*** 2) **AM/PM bug**: Flatpickr applies locale translation to AM/PM ONLY when date picker and timepicker are standalone components.
**_FIX_**: all non-english locales will default to 24H mode (no AM/PM) unless the `twentyFourHourFormat` boolean is explicitly set to true or false for those locales (otherwise undefined). for all english locales, it will default to 12H format
3) changed example hindi label to english to prevent confusion for devs so that they don't assume this value is being translated via the `locale` value. label is a string property and is to be set by developers, not shidoka's responsibility
4) added reset button to all 3 components (date picker, time picker, date range picker)
5) added multi-select date picker example on the `patterns/forms` page
6) corrected labels that read `Date / Time` to read `Date + Time`

attn: @brian-patrick-3  -- please see item 2 above and let me know your thoughts

great work and thanks for your help @srpriyankashetty !!

**Review by brian-patrick-3 — APPROVED** [review]

**Review by srpriyankashetty — APPROVED** [review]
1. Verified the new Date picker, Date range picker and Time picker changes and looks good
2. Verified that Required error message is displayed for all the pickers
3. Verified the min max range for all the pickers
4. Verified the range selections
5. Verified that Storybook >> Forms has all the above listed picker
6. Verified the date pickers and time pickers in Chrome, Edge and Firefox browser and looks fine
7. Verified the time for 12hrs and 24hrs format
8. Verified setting the locale and language changes in the calendar accordingly

Verified bugs list
Reset/clear button when a value is selected in all the pickers - Fixed
Multi select date picker is missing in the form page - Fixed
The label to 'Date / Time' must be changed to 'Date + Time' in form - Fixed
AM and PM is not changing based on the Locale set - set the time to 24hrs format
When there is Min and Max date range set, the year is not showing in the calendar. Also shows some arrows beside the month - Fixed

**Pull request merged**
